Launcher crashes after new phone updates

**Describe the crash**
After updating my phone to the most recent version of Android, the KISS launcher keeps crashing while navigating it. To the point I had to revert to the vanilla launcher of my phone.

**To Reproduce**
Ways in which I was able to crash the apk:
1. Moving back and forth between previously opened apps and home.
2. Touching the "close all apps" button.
3. Sometimes, although I could not catch the exception in this context, while searching the apps, contacts, etc.

**Stack trace**
If possible, please include a stack trace! You can get one:
```
I/ViewRootImpl@6605404[MainActivity]: Relayout returned: old=(0,0,1080,2280) new=(0,0,1080,2280) req=(1080,2280)8 dur=16 res=0x5 s={false 0} ch=true fn=2
I/ViewRootImpl@6605404[MainActivity]: stopped(true) old=false
D/DataHandler: Adding shortcut for com.Slack
D/DataHandler: Adding shortcut for com.Slack
D/DataHandler: Adding shortcut for com.Slack
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: fr.neamar.kiss.debug, PID: 10351
    java.lang.IllegalStateException: User 150 is locked or not running
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2445)
        at android.os.Parcel.createException(Parcel.java:2421)
        at android.os.Parcel.readException(Parcel.java:2404)
        at android.os.Parcel.readException(Parcel.java:2346)
        at android.content.pm.ILauncherApps$Stub$Proxy.getShortcuts(ILauncherApps.java:1573)
        at android.content.pm.LauncherApps.getShortcuts(LauncherApps.java:1169)
        at fr.neamar.kiss.utils.ShortcutUtil.getShortCut(ShortcutUtil.java:132)
        at fr.neamar.kiss.utils.ShortcutUtil.createShortcutRecord(ShortcutUtil.java:154)
        at fr.neamar.kiss.DataHandler.updateShortcut(DataHandler.java:535)
        at fr.neamar.kiss.DataHandler.updateShortcut(DataHandler.java:519)
        at fr.neamar.kiss.DataHandler.updateShortcuts(DataHandler.java:484)
        at fr.neamar.kiss.dataprovider.ShortcutsProvider$1.onShortcutsChanged(ShortcutsProvider.java:34)
        at android.content.pm.LauncherApps$CallbackMessageHandler.handleMessage(LauncherApps.java:1844)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:226)
        at android.os.Looper.loop(Looper.java:313)
        at android.app.ActivityThread.main(ActivityThread.java:8663)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:567)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.pm.ShortcutService.throwIfUserLockedL(ShortcutService.java:1329)
        at com.android.server.pm.ShortcutService$LocalService.getShortcuts(ShortcutService.java:3248)
        at com.android.server.pm.LauncherAppsService$LauncherAppsImpl.getShortcuts(LauncherAppsService.java:899)
        at android.content.pm.ILauncherApps$Stub.onTransact(ILauncherApps.java:682)
        at android.os.Binder.execTransactInternal(Binder.java:1215)
I/Process: Sending signal. PID: 10351 SIG: 9
Disconnected from the target VM, address: 'localhost:40787', transport: 'socket'
```

**Additional information**
 - App version: 3.16.10
 - Did the crash start happening with this version? No, it started after updating my phone.
 - App installed via: f-droid

**Additional context**
Add any other context about the problem here.
- Phone brand and model: Samsung S10e
- Base launcher: One UI 4.1
- Android version: 12
- Build number: SP1A.210812.016
- Kernel version: 4.14

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -507,8 +507,8 @@ public class DataHandler extends BroadcastReceiver
      * @param shortcut shortcut to be removed
      */
     public void removeShortcut(ShortcutPojo shortcut) {
-        removeShortcut(shortcut.id, shortcut.packageName, shortcut.intentUri);
-        if (this.getShortcutsProvider() != null) {
+        boolean shortcutUpdated = removeShortcut(shortcut.id, shortcut.packageName, shortcut.intentUri);
+        if (shortcutUpdated && this.getShortcutsProvider() != null) {
             this.getShortcutsProvider().reload();
         }
     }
@@ -564,8 +564,7 @@ public class DataHandler extends BroadcastReceiver
             return DBHelper.insertShortcut(this.context, shortcut);
         } else {
             String id = ShortcutUtil.generateShortcutId(shortcut.name);
-            removeShortcut(id, shortcut.packageName, shortcut.intentUri);
-            return true;
+            return removeShortcut(id, shortcut.packageName, shortcut.intentUri);
         }
     }
 
@@ -575,12 +574,13 @@ public class DataHandler extends BroadcastReceiver
      * @param id          KISS shortcut id, same as {@link ShortcutPojo#id}
      * @param packageName package name, same as {@link ShortcutPojo#packageName}
      * @param intentUri   intent to be called, same as {@link ShortcutPojo#intentUri}
+     * @return true, if shortcut was removed
      */
-    private void removeShortcut(String id, String packageName, String intentUri) {
+    private boolean removeShortcut(String id, String packageName, String intentUri) {
         Log.d(TAG, "Removing shortcut for " + packageName);
         // Also remove shortcut from favorites
         removeFromFavorites(id);
-        DBHelper.removeShortcut(this.context, packageName, intentUri);
+        return DBHelper.removeShortcut(this.context, packageName, intentUri);
     }
 
     /**

--- a/app/src/main/java/fr/neamar/kiss/DataHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/DataHandler.java
@@ -28,6 +28,7 @@ import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 
+import fr.neamar.kiss.broadcast.ProfileChangedHandler;
 import fr.neamar.kiss.dataprovider.AppProvider;
 import fr.neamar.kiss.dataprovider.ContactsProvider;
 import fr.neamar.kiss.dataprovider.IProvider;
@@ -85,6 +86,12 @@ public class DataHandler extends BroadcastReceiver
 
         Intent i = new Intent(MainActivity.START_LOAD);
         this.context.sendBroadcast(i);
+
+        // Monitor changes for profiles
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
+            ProfileChangedHandler profileChangedHandler = new ProfileChangedHandler();
+            profileChangedHandler.register(this.context.getApplicationContext());
+        }
 
         // Monitor changes for service preferences (to automatically start and stop services)
         SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);

--- a/app/src/main/java/fr/neamar/kiss/KissApplication.java
+++ b/app/src/main/java/fr/neamar/kiss/KissApplication.java
@@ -34,10 +34,6 @@ public class KissApplication extends Application {
         return dataHandler;
     }
 
-    public void setDataHandler(DataHandler newDataHandler) {
-        dataHandler = newDataHandler;
-    }
-
     public RootHandler getRootHandler() {
         if (rootHandler == null) {
             rootHandler = new RootHandler(this);

--- a/app/src/main/java/fr/neamar/kiss/broadcast/ProfileChangedHandler.java
+++ b/app/src/main/java/fr/neamar/kiss/broadcast/ProfileChangedHandler.java
@@ -1,0 +1,72 @@
+package fr.neamar.kiss.broadcast;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Build;
+import android.os.UserManager;
+
+import androidx.annotation.RequiresApi;
+
+import fr.neamar.kiss.DataHandler;
+import fr.neamar.kiss.KissApplication;
+import fr.neamar.kiss.dataprovider.AppProvider;
+import fr.neamar.kiss.dataprovider.ShortcutsProvider;
+import fr.neamar.kiss.utils.UserHandle;
+
+@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+public class ProfileChangedHandler extends BroadcastReceiver {
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+
+        if (Intent.ACTION_MANAGED_PROFILE_REMOVED.equals(intent.getAction())) {
+            // Try to clean up app-related data when profile is removed
+            android.os.UserHandle profile = intent.getParcelableExtra(Intent.EXTRA_USER);
+
+            // Package installation/uninstallation events for the main
+            // profile are still handled using PackageAddedRemovedHandler
+            UserManager manager = (UserManager) context.getSystemService(Context.USER_SERVICE);
+            UserHandle user = new UserHandle(manager.getSerialNumberForUser(profile), profile);
+
+            DataHandler dataHandler = KissApplication.getApplication(context).getDataHandler();
+            dataHandler.removeFromExcluded(user);
+            dataHandler.removeFromFavorites(user);
+        }
+
+        if (Intent.ACTION_MANAGED_PROFILE_ADDED.equals(intent.getAction()) ||
+                Intent.ACTION_MANAGED_PROFILE_REMOVED.equals(intent.getAction()) ||
+                Intent.ACTION_USER_UNLOCKED.equals(intent.getAction()) ||
+                Intent.ACTION_PROFILE_ACCESSIBLE.equals(intent.getAction()) ||
+                Intent.ACTION_PROFILE_INACCESSIBLE.equals(intent.getAction())) {
+            DataHandler dataHandler = KissApplication.getApplication(context).getDataHandler();
+
+            AppProvider appProvider = dataHandler.getAppProvider();
+            if (appProvider != null) {
+                appProvider.reload();
+            }
+            ShortcutsProvider shortcutsProvider = dataHandler.getShortcutsProvider();
+            if (shortcutsProvider != null) {
+                shortcutsProvider.reload();
+            }
+        }
+    }
+
+    public void register(Context context) {
+        IntentFilter filter = new IntentFilter();
+        filter.addAction(Intent.ACTION_MANAGED_PROFILE_ADDED);
+        filter.addAction(Intent.ACTION_MANAGED_PROFILE_REMOVED);
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            filter.addAction(Intent.ACTION_USER_UNLOCKED);
+        }
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            filter.addAction(Intent.ACTION_PROFILE_ACCESSIBLE);
+            filter.addAction(Intent.ACTION_PROFILE_INACCESSIBLE);
+        }
+
+        context.registerReceiver(this, filter);
+    }
+}

--- a/app/src/main/java/fr/neamar/kiss/dataprovider/AppProvider.java
+++ b/app/src/main/java/fr/neamar/kiss/dataprovider/AppProvider.java
@@ -1,6 +1,5 @@
 package fr.neamar.kiss.dataprovider;
 
-import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.Intent;
 import android.content.IntentFilter;
@@ -11,12 +10,8 @@ import android.os.Process;
 import android.os.UserManager;
 import android.preference.PreferenceManager;
 
-import androidx.annotation.RequiresApi;
-
 import java.util.ArrayList;
-import java.util.Objects;
 
-import fr.neamar.kiss.KissApplication;
 import fr.neamar.kiss.broadcast.PackageAddedRemovedHandler;
 import fr.neamar.kiss.loader.LoadAppPojos;
 import fr.neamar.kiss.normalizer.StringNormalizer;
@@ -30,7 +25,7 @@ public class AppProvider extends Provider<AppPojo> {
 
     @Override
     public void onCreate() {
-        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
             // Package installation/uninstallation events for the main
             // profile are still handled using PackageAddedRemovedHandler itself
             final UserManager manager = (UserManager) this.getSystemService(Context.USER_SERVICE);
@@ -90,28 +85,6 @@ public class AppProvider extends Provider<AppPojo> {
                     }
                 }
             });
-
-            // Try to clean up app-related data when profile is removed
-            IntentFilter filter = new IntentFilter();
-            filter.addAction(Intent.ACTION_MANAGED_PROFILE_ADDED);
-            filter.addAction(Intent.ACTION_MANAGED_PROFILE_REMOVED);
-            this.registerReceiver(new BroadcastReceiver() {
-                @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
-                @Override
-                public void onReceive(Context context, Intent intent) {
-                    if (Objects.equals(intent.getAction(), Intent.ACTION_MANAGED_PROFILE_ADDED)) {
-                        AppProvider.this.reload();
-                    } else if (Objects.equals(intent.getAction(), Intent.ACTION_MANAGED_PROFILE_REMOVED)) {
-                        android.os.UserHandle profile = intent.getParcelableExtra(Intent.EXTRA_USER);
-
-                        UserHandle user = new UserHandle(manager.getSerialNumberForUser(profile), profile);
-
-                        KissApplication.getApplication(context).getDataHandler().removeFromExcluded(user);
-                        KissApplication.getApplication(context).getDataHandler().removeFromFavorites(user);
-                        AppProvider.this.reload();
-                    }
-                }
-            }, filter);
         }
 
         // Get notified when app changes on standard user profile

--- a/app/src/main/java/fr/neamar/kiss/db/DBHelper.java
+++ b/app/src/main/java/fr/neamar/kiss/db/DBHelper.java
@@ -225,12 +225,19 @@ public class DBHelper {
         return records;
     }
 
+    /**
+     * Insert or update a shortcut into DB.
+     *
+     * @param context
+     * @param shortcut
+     * @return true, if shortcut has changed
+     */
     public static boolean insertShortcut(Context context, ShortcutRecord shortcut) {
         SQLiteDatabase db = getDatabase(context);
-        // Do not add duplicate shortcuts
-        try (Cursor cursor = db.query("shortcuts", new String[]{"package", "intent_uri"},
-                "package = ? AND intent_uri = ?", new String[]{shortcut.packageName, shortcut.intentUri}, null, null, null, null)) {
-            if (cursor.moveToFirst()) {
+        // check if any field has changed
+        try (Cursor cursor = db.query("shortcuts", new String[]{"name", "package", "intent_uri"},
+                "name = ? and package = ? AND intent_uri = ?", new String[]{shortcut.name, shortcut.packageName, shortcut.intentUri}, null, null, null, null)) {
+            if (cursor.getCount() > 0) {
                 return false;
             }
         }
@@ -242,13 +249,26 @@ public class DBHelper {
         values.put("icon_blob", (String) null); // Another legacy field (icon is dynamically retrieved)
         values.put("intent_uri", shortcut.intentUri);
 
-        db.insert("shortcuts", null, values);
+        // do not add duplicate shortcuts
+        int rowsAffected = db.update("shortcuts", values, "package = ? AND intent_uri = ?", new String[]{shortcut.packageName, shortcut.intentUri});
+        if (rowsAffected == 0) {
+            db.insert("shortcuts", null, values);
+        }
         return true;
     }
 
-    public static void removeShortcut(Context context, String packageName, String intentUri) {
+    /**
+     * Remove a shortcut from DB.
+     *
+     * @param context
+     * @param packageName
+     * @param intentUri
+     * @return true, if shortcut was removed
+     */
+    public static boolean removeShortcut(Context context, String packageName, String intentUri) {
         SQLiteDatabase db = getDatabase(context);
-        db.delete("shortcuts", "package = ? AND intent_uri = ?", new String[]{packageName, intentUri});
+        int rowsAffected = db.delete("shortcuts", "package = ? AND intent_uri = ?", new String[]{packageName, intentUri});
+        return rowsAffected > 0;
     }
 
     public static void addCustomAppName(Context context, String componentName, String newName) {

--- a/app/src/main/java/fr/neamar/kiss/loader/LoadShortcutsPojos.java
+++ b/app/src/main/java/fr/neamar/kiss/loader/LoadShortcutsPojos.java
@@ -1,6 +1,8 @@
 package fr.neamar.kiss.loader;
 
 import android.content.Context;
+import android.content.pm.ShortcutInfo;
+import android.os.Build;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -37,10 +39,22 @@ public class LoadShortcutsPojos extends LoadPojos<ShortcutPojo> {
             pojo.setName(shortcutRecord.name);
             pojo.setTags(tagsHandler.getTags(pojo.id));
 
-            pojos.add(pojo);
+            if (isExistingShortcut(pojo)) {
+                pojos.add(pojo);
+            }
         }
 
         return pojos;
+    }
+
+    private boolean isExistingShortcut(ShortcutPojo pojo) {
+        if (pojo.isOreoShortcut()) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                ShortcutInfo shortcutInfo = ShortcutUtil.getShortCut(context.get(), pojo.packageName, pojo.getOreoId());
+                return shortcutInfo != null;
+            }
+        }
+        return true;
     }
 
 }

--- a/app/src/main/java/fr/neamar/kiss/utils/ShortcutUtil.java
+++ b/app/src/main/java/fr/neamar/kiss/utils/ShortcutUtil.java
@@ -128,7 +128,7 @@ public class ShortcutUtil {
 
             // find the correct UserHandle and get shortcut
             for (UserHandle userHandle : userHandles) {
-                if (userManager.isUserRunning(userHandle)) {
+                if (userManager.isUserRunning(userHandle) && userManager.isUserUnlocked(userHandle)) {
                     List<ShortcutInfo> shortcuts = launcherApps.getShortcuts(query, userHandle);
                     if (shortcuts != null) {
                         for (ShortcutInfo shortcut : shortcuts) {

